[BUG] In workspaces `npm run` autocomplete only lists root scripts

### Is there an existing issue for this?

- [X] I have searched the existing issues

### This issue exists in the latest npm version

- [X] I am using the latest npm

### Current Behavior

When run from a workspace sub-folder, the `npm run` autocomplete only lists scripts from the workspace root package.json. This is a problem because those aren't the scripts that `npm run` actually can execute in that context. Trying to run one of the autocompleted script names will fail with "npm ERR! Error: Missing script".

### Expected Behavior

It should list scripts from the nearest ancestor package.json, as that is what will actually be executed, not the ones from the root.

### Steps To Reproduce

1. Make a new npm project with at least one workspace
2. Add a script to the root package.json file
3. Add a script to the workspace's package.json file
4. `cd` to the workspace folder
5. Type `npm run` and press <kbd>Tab</kbd>

### Environment

- npm: 10.2.5 (also reproducible on 9.8.1)
- Node.js: 20.10.0 (also reproducible on 18.18.2)
- OS Name: macOS
- System Model Name: Macbook Pro
- npm config:
```ini
npm ERR! code ENOWORKSPACES
npm ERR! This command does not support workspaces.
```

--- a/lib/commands/run-script.js
+++ b/lib/commands/run-script.js
@@ -26,7 +26,11 @@ class RunScript extends BaseCommand {
   static async completion (opts, npm) {
     const argv = opts.conf.argv.remain
     if (argv.length === 2) {
-      const { content: { scripts = {} } } = await pkgJson.normalize(npm.localPrefix)
+      const workspacePrefixes = npm.config.get('workspace', 'default')
+      const localPrefix = workspacePrefixes.length
+        ? workspacePrefixes[0]
+        : npm.localPrefix
+      const { content: { scripts = {} } } = await pkgJson.normalize(localPrefix)
         .catch(() => ({ content: {} }))
       if (opts.isFish) {
         return Object.keys(scripts).map(s => `${s}\t${scripts[s].slice(0, 30)}`)

--- a/test/lib/commands/run-script.js
+++ b/test/lib/commands/run-script.js
@@ -3,6 +3,7 @@ const { resolve } = require('node:path')
 const realRunScript = require('@npmcli/run-script')
 const mockNpm = require('../../fixtures/mock-npm')
 const { cleanCwd } = require('../../fixtures/clean-snapshot')
+const path = require('node:path')
 
 const mockRs = async (t, { windows = false, runScript, ...opts } = {}) => {
   let RUN_SCRIPTS = []
@@ -491,6 +492,7 @@ t.test('workspaces', async t => {
     prefixDir,
     workspaces = true,
     exec = [],
+    chdir = ({ prefix }) => prefix,
     ...config
   } = {}) => {
     const mock = await mockRs(t, {
@@ -554,6 +556,7 @@ t.test('workspaces', async t => {
         ...Array.isArray(workspaces) ? { workspace: workspaces } : { workspaces },
         ...config,
       },
+      chdir,
       runScript,
     })
     if (exec) {
@@ -561,6 +564,22 @@ t.test('workspaces', async t => {
     }
     return mock
   }
+
+  t.test('completion', async t => {
+    t.test('in root dir', async t => {
+      const { runScript } = await mockWorkspaces(t)
+      const res = await runScript.completion({ conf: { argv: { remain: ['npm', 'run'] } } })
+      t.strictSame(res, [])
+    })
+
+    t.test('in workspace dir', async t => {
+      const { runScript } = await mockWorkspaces(t, {
+        chdir: ({ prefix }) => path.join(prefix, 'packages/c'),
+      })
+      const res = await runScript.completion({ conf: { argv: { remain: ['npm', 'run'] } } })
+      t.strictSame(res, ['test', 'posttest', 'lorem'])
+    })
+  })
 
   t.test('list all scripts', async t => {
     const { joinedOutput } = await mockWorkspaces(t)

